EIPS/eip-5216.md

```diff
@@ -0,0 +1,116 @@
+discussions-to: https://ethereum-magicians.org/t/eip-erc1155-approval-by-amount/9898
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-07-11
+requires: 20, 1155
+---
+
+## Abstract
+
+This specification defines standard functions for approving amounts of an [ERC-1155](./eip-1155.md). An implementation allows approve by id, different amounts both batch and individual. The proposal depends on and extends the existing [ERC-1155](./eip-1155.md).
+
+## Motivation
+
+[ERC-1155](./eip-1155.md) has created a scenario where multiple token management and transactions occur on a daily basis. Although it can be used as [ERC-721](./eip-721.md), the most common way to use it is to create collections through its `ids` in order to have an indefinite `amount` of tokens of one type. These tokens have been implemented in a wide variety of projects, but the most important are marketplaces.
+The nature of tokens is trading, transfer and use, so it is important that your peer-to-peer transactions are as secure as possible.
+
+The way the [ERC-1155](./eip-1155.md) standard approves tokens is with the `setApprovalForAll(address operator, bool approved)` function, which approves ALL tokens of an `id`. However, this is a standard that combines ideas from the [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) standards, and it is important to create a trust mechanism, where an owner can allow a third party, such as a marketplace, to approve a limited number of tokens of an `id` and not all at once.
+
+In turn, it is also necessary to subtract the amount of tokens approved once they have been successfully transferred to avoid the scenario that we have now, where ALL tokens of an `id` are approved without revoking that approval at any time. In our honest opinion, this is quite dangerous because if you don't revoke that approval with `setApprovalForAll(operatorAddress, false)`, you let that the operator do whateaver he wants whenever he wants with all your tokens.
+
+By having a way to approve and revoke as the [ERC-20](./eip-20.md) standard works, i.e. by amounts we reduce that security problem:
+
+- By `approve` function, we allow an operator to use an amount of tokens per id.
+- Through the function `allowance` you can see the amount of tokens that an approved account has per id.
+
+Both following the [ERC-20](./eip-20.md) standard name patterns.
+
+
+## Specification
+
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+Every contract compliant to the `ERC1155ApprovalByAmount` extension MUST implement the `IERC1155ApprovalByAmount` interface. The **approval by amount extension** is OPTIONAL for [ERC-1155](./eip-1155.md) contracts.
+
+### Interface implementation
+
+```solidity
+/**
+ * @title ERC-1155 Approval By Amount Extension
+ * Note: the ERC-165 identifier for this interface is 0x1be07d74
```

Comment:
You'll want to include 165 in your requires header.

Reply:
Included!